lightning/src/util/ser_macros.rs

```diff
@@ -135,12 +145,21 @@ macro_rules! impl_writeable {
 	}
 }
 macro_rules! impl_writeable_len_match {
-	($st:ident, {$({$m: pat, $l: expr}),*}, {$($field:ident),*}) => {
+	($st:ident, $cmp: tt, {$({$m: pat, $l: expr}),*}, {$($field:ident),*}) => {
 		impl Writeable for $st {
 			fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
-				w.size_hint(match *self {
+				let len = match *self {
 					$($m => $l,)*
-				});
+				};
+				w.size_hint(len);
+				#[cfg(any(test, feature = "fuzztarget"))]
+				{
+					// In tests, assert that the hard-coded length matches the actual one
+					use util::ser::LengthCalculatingWriter;
+					let mut len_calc = LengthCalculatingWriter(0);
+					$( self.$field.write(&mut len_calc)?; )*
+					assert!(len_calc.0 $cmp len);
```

Comment:
Wondering the reason why we supply the length to begin with at all, and not just use `LengthCalculatingWriter` from the get-go?

Reply:
In most cases that would be fine, in some (like `NodeAnnouncement`) we'd rather not have to iterate over some inner guts to figure out the length just for a hint. More importantly, I'm not exactly sure how well LLVM optimizes passes like that - there's a lot of Errs and unwraps in serialization stuff, so you have to get LLVM to aggressively inline for it to be reasonable, and I'm not sure that it does.